diff --git a/models/image_recognition/tensorflow/resnet50v1_5/inference/gpu/fp16/preprocessing.py b/models/image_recognition/tensorflow/resnet50v1_5/inference/gpu/fp16/preprocessing.py
index bf81b418..1b774fbd 100644
--- a/models/image_recognition/tensorflow/resnet50v1_5/inference/gpu/fp16/preprocessing.py
+++ b/models/image_recognition/tensorflow/resnet50v1_5/inference/gpu/fp16/preprocessing.py
@@ -133,7 +133,7 @@ class RecordInputImagePreprocessor(object):
     image = tf.image.decode_jpeg(
       image_buffer, channels=3, fancy_upscaling=False, dct_method='INTEGER_FAST')
     image = eval_image(image, self.height, self.width, self.resize_method)
-    image = tf.transpose(image, perm=[2, 0, 1])
+    #image = tf.transpose(image, perm=[2, 0, 1])
     return (image, label_index, filename)
 
   def minibatch(self, dataset, subset, cache_data=False):
